controllers/helmchart_controller.go

```diff
@@ -426,16 +435,147 @@ func (r *HelmChartReconciler) reconcileFromTarballArtifact(ctx context.Context,
+			}
+
+			if _, err = os.Stat(origPath); os.IsNotExist(err) {
+				err := fmt.Errorf("directory %s not found: %s", origPath, err)
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			} else if err != nil {
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			}
+
+			ch, err := loader.Load(origPath)
+			if err != nil {
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			}
+
+			constraint, err := semver.NewConstraint(dep.Version)
+			if err != nil {
+				err := fmt.Errorf("dependency %s has an invalid version/constraint format: %s", dep.Name, err)
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			}
+
+			v, err := semver.NewVersion(ch.Metadata.Version)
+			if err != nil {
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			}
+
+			if constraint.Check(v) {
+				_, err = chartutil.Save(ch, destPath)
+				if err != nil {
+					return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+				}
+			} else {
+				err := fmt.Errorf("can't get a valid version for dependency %s.", dep.Name)
+				return sourcev1.HelmChartNotReady(chart, sourcev1.StorageOperationFailedReason, err.Error()), err
+			}
+		} else {
+			// Discover existing HelmRepository by URL
+			repo, err := r.resolveDependencyRepo(dep, chart.Namespace)
+			var clientOpts []getter.Option
+
+			if repo != nil && repo.Spec.SecretRef != nil {
+				name := types.NamespacedName{
+					Namespace: repo.GetNamespace(),
+					Name:      repo.Spec.SecretRef.Name,
+				}
+
+				var secret corev1.Secret
+				err := r.Client.Get(ctx, name, &secret)
+				if err != nil {
+					err = fmt.Errorf("auth secret error: %w", err)
+					return sourcev1.HelmChartNotReady(chart, sourcev1.AuthenticationFailedReason, err.Error()), err
+				}
+
+				opts, cleanup, err := helm.ClientOptionsFromSecret(secret)
+				if err != nil {
+					err = fmt.Errorf("auth options error: %w", err)
+					return sourcev1.HelmChartNotReady(chart, sourcev1.AuthenticationFailedReason, err.Error()), err
+				}
+				defer cleanup()
+				clientOpts = append(opts, getter.WithTimeout(repo.GetTimeout()))
+			}
+
+			// Initialize the chart repository and load the index file
+			chartRepo, err := helm.NewChartRepository(dep.Repository, r.Getters, clientOpts)
+			if err != nil {
+				switch err.(type) {
+				case *url.Error:
+					return sourcev1.HelmChartNotReady(chart, sourcev1.URLInvalidReason, err.Error()), err
+				default:
+					return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+				}
+			}
+
+			// Download index
+			err = chartRepo.DownloadIndex()
+			if err != nil {
+				return sourcev1.HelmChartNotReady(chart, sourcev1.ChartPullFailedReason, err.Error()), err
+			}
```

Comment:
Instead of always downloading the index, we should either retrieve it from the artifact that is attached to the `HelmRelease` resource and load it using `ChartRepository#LoadIndex`, or download the artifact if we do not have a resource for the repository URL.

This makes the artifact function as a cache.